pkg/apiserver/registry/core/shoot/strategy.go

```diff
@@ -190,6 +191,15 @@ func (shootStrategy) Canonicalize(obj runtime.Object) {
 		shoot.Spec.Kubernetes.ClusterAutoscaler.MaxEmptyBulkDelete = nil
 	}
 	gardenerutils.MaintainSeedNameLabels(shoot, shoot.Spec.SeedName, shoot.Status.SeedName)
+	maintainIsSelfHostedLabel(shoot)
+}
+
+func maintainIsSelfHostedLabel(shoot *core.Shoot) {
+	if !gardencorehelper.IsShootSelfHosted(shoot.Spec.Provider.Workers) {
+		delete(shoot.Labels, v1beta1constants.ShootIsSelfHosted)
+	} else {
+		metav1.SetMetaDataLabel(&shoot.ObjectMeta, v1beta1constants.ShootIsSelfHosted, "true")
+	}
```

Comment:
Nit: As a positive person, I prefer less negation. Can we switch the order? From my point of view it is easier to understand without a negation.
```suggestion
	if gardencorehelper.IsShootSelfHosted(shoot.Spec.Provider.Workers) {
		metav1.SetMetaDataLabel(&shoot.ObjectMeta, v1beta1constants.ShootIsSelfHosted, "true")
	} else {
		delete(shoot.Labels, v1beta1constants.ShootIsSelfHosted)
	}
```